### PR TITLE
chore: upgrade phpstan to v2

### PR DESCRIPTION
**Pull request by blackshadev**
# What

- Upgrade phpstan to v2
- Upgrade required dependencies to other versions
  - rector v2.0-rc2
  - spaze/phpstan-disallowed-calls v4
- Remove phpat, it doesn't yet support phpstan, and I cannot see any progress on it. I have found [this issue](https://github.com/carlosas/phpat/issues/293) , but no branches revering it etc. 

## Why not wait?

Well , as it stands, phpstan doesn't allot for property hooks, I tried a property hook in an interface and phpstan paniced to the point it became unusable. In my opinion, I rather have php 8.4 features, than the features added by phpat, which can be re-added when phpat does provide phpstan 2.0 support.

## Why the baseline

I rather introduce phpstan 2.0 early and fix the issues from the baseline later, the alternative is my trying to clean up while others keep on piling new issues without them even knowing they did it (because phpstan 2 has become more strict)

### CODE REVIEW

File: phpstan-baseline.neon
Comment (aidan-casey) [inline phpstan-baseline.neon:1]: @blackshadev can you give me the cliff notes on this one? I'd rather avoid ignoring stuff if we can, but it kind of looks like some of these are related to PHPUnit?

File: tests/Architecture/ArchitectureTestCase.php
Comment (aidan-casey) [inline tests/Architecture/ArchitectureTestCase.php:1]: We can leave this file in place without it causing issues, right? Seems like we'd want to add phpat in the future, so we should just add a comment and leave it.

**Review by aidan-casey — COMMENTED** [review]
@blackshadev overall looks good, thank you! Just a few comments.

File: tests/Architecture/ArchitectureTestCase.php
Comment (brendt) [reply tests/Architecture/ArchitectureTestCase.php:1]: We could also restore it back from GIT if we need it. I rather have the current codebase as clean as possible :)

**Comment by brendt** [comment]
Nice!

File: phpstan-baseline.neon
Comment (blackshadev) [reply phpstan-baseline.neon:1]: This file is generated using `vendor/bin/phpstan  analyse src tests --memory-limit=1G --generate-baseline` .
What is does, it automatically creates a rule in the baseline to ignore the specific error it finds. 

This means phpstan wil only report on newly found error and not errors already in the codebase. 

Of course in the long run we can resolve these issues and thus remove the baseline. But doing that in one PR is "dwijlen met de kraan open", new people will merge new code, which I need to rebase on, which introduces new errors which phpstan 2 finds but the people introducing them didn't see. I fix error and new code is merged to main, at infinitum.

File: tests/Architecture/ArchitectureTestCase.php
Comment (aidan-casey) [reply tests/Architecture/ArchitectureTestCase.php:1]: @brendt I'm going to completely forget about it if we do that. 😅

File: phpstan-baseline.neon
Comment (aidan-casey) [reply phpstan-baseline.neon:1]: Thank you, and understood. 🙂

File: tests/Architecture/ArchitectureTestCase.php
Comment (brendt) [reply tests/Architecture/ArchitectureTestCase.php:1]: I won't forget ;)

**Comment by blackshadev** [comment]
Pipeline fixed, newly introduced code from main had a new error in it, adding it to the baseline for now.

**Comment by brendt** [comment]
Let's go!

**Comment by carlosas** [comment]
FYI @blackshadev @aidan-casey @brendt 
PHPat now supports PHPStan 2.0 https://github.com/carlosas/phpat/releases/tag/0.11.0

**Comment by aidan-casey** [comment]
Thank you, @carlosas! Appreciate your continued work on the project!

**Comment by blackshadev** [comment]
> FYI @blackshadev @aidan-casey @brendt PHPat now supports PHPStan 2.0 https://github.com/carlosas/phpat/releases/tag/0.11.0 Support for PHP 8.4 (like catching dependencies in property hooks) is WIP

Thanks, created an MR to re-add it https://github.com/tempestphp/tempest-framework/pull/873

**Pull request merged**